### PR TITLE
Addresses #3906, support new field 'Minimum Air Flow Turndown Schedule Name' for AirTerminals

### LINKED ISSUE #3906: E+ 9.3.0: Support new field 'Minimum Air Flow Turndown Schedule Name' for AirTerminals
Enhancement Request
--------------

* ATU got a new field at end `Minimum Air Flow Turndown Schedule Name` : 
    * `AirTerminal:SingleDuct:VAV:HeatAndCool:NoReheat`, `AirTerminal:SingleDuct:ConstantVolume:Reheat`, `AirTerminal:SingleDuct:VAV:Reheat` , `AirTerminal:SingleDuct:VAV:HeatAndCool:Reheat`, `AirTerminal:DualDuct:VAV`
(`AirTerminal:SingleDuct:VAV:Reheat:VariableSpeedFan`)
* `AirTerminal:SingleDuct:Uncontrolled`, which was obsolete, is removed. This will not affect us as

### PR DESCRIPTION
**Pull request by joseph-robertson**
Pull request overview
---------------------

 - Fixes #3906
 - Adds FT tests for:
   - `AirTerminal:SingleDuct:VAV:Reheat`
   - `AirTerminal:SingleDuct:VAV:NoReheat`
   - `AirTerminal:SingleDuct:VAV:HeatAndCool:Reheat`
   - `AirTerminal:SingleDuct:VAV:HeatAndCool:NoReheat`
   - `AirTerminal:DualDuct:VAV` 

Questions:
* Do we need version translation if we're just adding a new optional field? Answer: since the new field is at the end, we don't need it.

### Pull Request Author

 - [x] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [x] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: resources/model/OpenStudio.idd [hunk 1]
```diff
@@ -10971,10 +10971,19 @@ OS:AirTerminal:SingleDuct:VAV:HeatAndCool:NoReheat,
...
        \maximum 1.0
+  A6;  \field Minimum Air Flow Turndown Schedule Name
+       \type object-list
+       \object-list ScheduleNames
+       \note This field adjusts the design minimum flow rate by multiplying it using this schedule of fraction
+       \note values. This field is used with "Zone Minimum Air Flow Fraction". Schedule values are fractions
+       \note 0.0 to 1.0. This field adjusts the minimum airflow turndown below the design minimum air flow
+       \note and is intended for use with ASHRAE Standard 170. If this field is left blank, then the turndown
+       \note minimum air flow fraction value is set to 1 and the model uses the fixed fraction specified in
+       \note in the field "Zone Minimum Air Flow Fraction".
```
Comment (jmarrec) [inline resources/model/OpenStudio.idd:10986]: Have you considered making this a required-field and setting to alwaysOnDiscreteSchedule instead? (this comment applies to all of the objects)

File: resources/model/OpenStudio.idd [hunk 2]
```diff
@@ -11857,6 +11891,15 @@ OS:AirTerminal:DualDuct:VAV,
...
        \note If this field is blank, then the terminal unit will not be controlled for outdoor air flow.
+  A8 ; \field Minimum Air Flow Turndown Schedule Name
+       \type object-list
+       \object-list ScheduleNames
+       \note This field adjusts the design minimum flow rate by multiplying it using this schedule of fraction
+       \note values. This field is used with "Zone Minimum Air Flow Fraction". Schedule values are fractions
+       \note 0.0 to 1.0. This field adjusts the minimum airflow turndown below the design minimum air flow
+       \note and is intended for use with ASHRAE Standard 170. If this field is left blank, then the turndown
+       \note minimum air flow fraction value is set to 1 and the VAV air terminal uses fixed fraction specified
+       \note in the field "Zone Minimum Air Flow Fraction".
```
Comment (jmarrec) [inline resources/model/OpenStudio.idd:11902]: Ok, so this is the last field added in all objects, so unless it's required-field, no VT is expected to happen.

File: src/energyplus/ForwardTranslator/ForwardTranslateAirTerminalDualDuctVAV.cpp [hunk 3]
```diff
@@ -101,6 +101,17 @@ namespace energyplus {
...
     }
 
+    // MinimumAirFlowTurndownScheduleName
+    boost::optional<Schedule> minAirFlowTurndownSchedule = modelObject.minimumAirFlowTurndownSchedule();
+
+    if (minAirFlowTurndownSchedule) {
+      boost::optional<IdfObject> _minAirFlowTurndownSchedule = translateAndMapModelObject(minAirFlowTurndownSchedule.get());
+
+      if (_minAirFlowTurndownSchedule && _minAirFlowTurndownSchedule->name()) {
+        idfObject.setString(AirTerminal_DualDuct_VAVFields::MinimumAirFlowTurndownScheduleName, _minAirFlowTurndownSchedule->name().get());
```
Comment (jmarrec) [inline src/energyplus/ForwardTranslator/ForwardTranslateAirTerminalDualDuctVAV.cpp:111]: I'd prefer doing this, but this is kind of a nitpick. if the schedule is translated, we KNOW it has a name so the check is redundant and (minimally) hurting performance

```suggestion
      if (_minAirFlowTurndownSchedule) {
        idfObject.setString(AirTerminal_DualDuct_VAVFields::MinimumAirFlowTurndownScheduleName, _minAirFlowTurndownSchedule->nameString());
```

File: src/model/AirTerminalDualDuctVAV.cpp [hunk 4]
```diff
@@ -159,6 +173,32 @@ namespace model {
       OS_ASSERT(result);
     }
 
+    bool AirTerminalDualDuctVAV_Impl::setMinimumAirFlowTurndownSchedule(Schedule& schedule) {
+      bool result = setSchedule(OS_AirTerminal_DualDuct_VAVFields::MinimumAirFlowTurndownScheduleName, "AirTerminalSingleDuctVAVReheat",
+                                "Minimum Air Flow Turndown", schedule);
+      return result;
+    }
```
Comment (jmarrec) [inline src/model/AirTerminalDualDuctVAV.cpp:180]: * [ ] TODO: For this file, I believe you forgot to modify getScheduleTypeKeys. on line 79

File: src/model/AirTerminalSingleDuctVAVHeatAndCoolNoReheat.cpp [hunk 5]
```diff
@@ -91,6 +91,9 @@ namespace model {
       if (std::find(b, e, OS_AirTerminal_SingleDuct_VAV_HeatAndCool_NoReheatFields::AvailabilitySchedule) != e) {
         result.push_back(ScheduleTypeKey("AirTerminalSingleDuctVAVHeatAndCoolNoReheat", "Availability Schedule"));
       }
+      if (std::find(b, e, OS_AirTerminal_SingleDuct_VAV_HeatAndCool_NoReheatFields::MinimumAirFlowTurndownScheduleName) != e) {
+        result.push_back(ScheduleTypeKey("AirTerminalSingleDuctVAVReheat", "Minimum Air Flow Turndown"));
+      }
```
Comment (jmarrec) [inline src/model/AirTerminalSingleDuctVAVHeatAndCoolNoReheat.cpp:96]: Ok, getScheduleTypeKeys is correctly modified for this file.

File: src/model/ScheduleTypeRegistry.cpp [hunk 6]
```diff
@@ -131,6 +131,7 @@ namespace model {
       {"AirLoopHVACUnitarySystem", "Supply Air Fan Operating Mode", "supplyAirFanOperatingModeSchedule", false, "ControlMode", 0.0, 1.0},
       {"AirTerminalDualDuctConstantVolume", "Availability Schedule", "availabilitySchedule", false, "Availability", 0.0, 1.0},
       {"AirTerminalDualDuctVAV", "Availability Schedule", "availabilitySchedule", false, "Availability", 0.0, 1.0},
+      {"AirTerminalDualDuctVAV", "Minimum Air Flow Turndown", "minimumAirFlowTurndown", true, "Availability", 0.0, 1.0},
```
Comment (jmarrec) [inline src/model/ScheduleTypeRegistry.cpp:134]: ScheduleTypeRegistry is correctly adding it, but getScheduleTypeKeys wasn't.

Comment (jmarrec) [reply src/model/ScheduleTypeRegistry.cpp:134]: rest of file is ok, I see 5 additions as expected.

File: src/model/test/AirTerminalDualDuctVAV_GTest.cpp [hunk 7]
```diff
@@ -135,3 +138,22 @@ TEST_F(ModelFixture, AirTerminalDualDuctVAV) {
     }
   }
 }
+
+TEST_F(ModelFixture, AirTerminalDualDuctVAV_MinimumAirFlowTurndownSchedule) {
```
Comment (jmarrec) [inline src/model/test/AirTerminalDualDuctVAV_GTest.cpp:142]: Ok, I see 5 model tests. 

* [ ] TODO: I would like to see 5 FT tests too (or modification of existing FT tests) for the new fields.

**Review by jmarrec — CHANGES_REQUESTED** [review]
Ok with all changes, they look good.

Missing right now:
* [x] AirTerminalDualDuctVAV::getScheduleTypeKeys
* [x] FT tests
* ~[ ] perhaps a new regression test?~ Added bunch of FT tests instead.

I wonder if this shouldn't be a required field defaulting to alwaysOnDiscreteSchedule.

File: src/energyplus/ForwardTranslator/ForwardTranslateAirTerminalDualDuctVAV.cpp [hunk 8]
```diff
@@ -101,6 +101,17 @@ namespace energyplus {
       }
     }
 
+    // MinimumAirFlowTurndownScheduleName
+    boost::optional<Schedule> minAirFlowTurndownSchedule = modelObject.minimumAirFlowTurndownSchedule();
+
+    if (minAirFlowTurndownSchedule) {
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/energyplus/ForwardTranslator/ForwardTranslateAirTerminalDualDuctVAV.cpp:107]: (same below, for translateAndMapModelObject, variable declaration can be made inside `if`)

File: src/energyplus/CMakeLists.txt [hunk 9]
```diff
@@ -605,9 +605,14 @@ set(${target_name}_test_src
...
   Test/AirLoopHVACUnitaryHeatCoolVAVChangeoverBypass_GTest.cpp
   Test/AirTerminalDualDuctConstantVolume_GTest.cpp
+  Test/AirTerminalDualDuctVAV_GTest.cpp
   Test/AirTerminalDualDuctVAVOutdoorAir_GTest.cpp
   Test/AirTerminalSingleDuctConstantVolumeFourPipeBeam_GTest.cpp
   Test/AirTerminalSingleDuctInletSideMixer_GTest.cpp
+  Test/AirTerminalSingleDuctVAVNoReheat_Gtest.cpp
+  Test/AirTerminalSingleDuctVAVReheat_Gtest.cpp
+  Test/AirTerminalSingleDuctVAVHeatAndCoolNoReheat_Gtest.cpp
+  Test/AirTerminalSingleDuctVAVHeatAndCoolReheat_Gtest.cpp
```
Comment (jmarrec) [inline src/energyplus/CMakeLists.txt:615]: @joseph-robertson be careful: windows isn't case sensitive but Ubuntu is: `Gtest` != `GTest`

```diff
-  Test/AirTerminalSingleDuctVAVHeatAndCoolNoReheat_Gtest.cpp
+  Test/AirTerminalSingleDuctVAVHeatAndCoolNoReheat_GTest.cpp
```

This is why the Ubuntu runners are failing at CMake configure: https://ci.commercialbuildings.dev/blue/organizations/jenkins/openstudio-incremental-ubuntu-18-04/detail/PR-4381/18/pipeline#step-37-log-253

Comment (jmarrec) [reply src/energyplus/CMakeLists.txt]: Addressed in aa348bb1dd54efb4ebdba0ebdf72fa02fc0b0871

File: resources/model/OpenStudio.idd (hunk 1 above)
Comment (joseph-robertson) [reply resources/model/OpenStudio.idd:10986]: For consistency, I think we should follow the other schedule fields (i.e., not making it a required-field and setting to alwaysOnDiscreteSchedule).

**Review by jmarrec — CHANGES_REQUESTED** [review]
ScheduleTypeRegistry change is needed. Otherwise this is great, and thanks for the effort creating FT tests for these objects from scratch! :+1: @joseph-robertson

File: src/model/ScheduleTypeRegistry.cpp [hunk 10]
```diff
@@ -144,10 +145,14 @@ namespace model {
       {"AirTerminalSingleDuctConstantVolumeNoReheat", "Availability", "availabilitySchedule", false, "Availability", 0.0, 1.0},
       {"AirTerminalSingleDuctVAVReheat", "Availability", "availabilitySchedule", false, "Availability", 0.0, 1.0},
       {"AirTerminalSingleDuctVAVReheat", "Minimum Air Flow Fraction", "minimumAirFlowFractionSchedule", true, "", 0.0, 1.0},
+      {"AirTerminalSingleDuctVAVReheat", "Minimum Air Flow Turndown", "minimumAirFlowTurndownSchedule", true, "Availability", 0.0, 1.0},
```
[the remark this answers is not in the document]
Comment (joseph-robertson) [reply src/model/ScheduleTypeRegistry.cpp:148]: Good catch.

**Review by jmarrec — APPROVED** [review]

**Pull request merged**